android/src/main/kotlin/com.netguru.multiplatformstorage/AndroidStorageProvider.kt

```diff
@@ -25,19 +52,123 @@ internal object AndroidStorageProvider {
      * In other cases, it uses default instance.
      */
     internal fun preferences(name: String? = null): Preferences {
+        // Generate a new encryption key with initialization vector.
+        val keyStore = KeyStore.getInstance(KEYSTORE_NAME)
+        keyStore.load(null)
+
+        val preferences = appContext.getSharedPreferences(PREFERENCES_NAME, Context.MODE_PRIVATE)
+
+        // Generating new keys
+        if (!keysExist(keyStore, preferences)) {
+            clearKeys(keyStore, preferences)
+            generateKeystoreKey(KEYSTORE_VALUES_KEY_ALIAS, keyStore)
+            generateKeystoreKey(KEYSTORE_VALUES_IV_ALIAS, keyStore)
+            generateKeystoreKey(KEYSTORE_KEYS_KEY_ALIAS, keyStore)
+            preferences.edit {
+                putString(VALUES_KEY_ALIAS, generateEncryptedKey(keyStore.getSecretKey(KEYSTORE_VALUES_KEY_ALIAS)))
+                putString(VALUES_IV_ALIAS, generateEncryptedKey(keyStore.getSecretKey(KEYSTORE_VALUES_IV_ALIAS)))
+                putString(KEYS_KEY_ALIAS, generateEncryptedKey(keyStore.getSecretKey(KEYSTORE_KEYS_KEY_ALIAS)))
+            }
+        }
+
+        val valuesKey = getDecryptedKey(VALUES_KEY_ALIAS, KEYSTORE_VALUES_KEY_ALIAS, preferences, keyStore)
+        val valuesIv = getDecryptedKey(VALUES_IV_ALIAS, KEYSTORE_VALUES_IV_ALIAS, preferences, keyStore)
+        val keysKey = getDecryptedKey(KEYS_KEY_ALIAS, KEYSTORE_KEYS_KEY_ALIAS, preferences, keyStore)
+
+
+        return BinaryPreferencesBuilder(appContext)
+            .migrateFrom(getSharedPreferences(name))
+            .keyEncryption(XorKeyEncryption(keysKey))
```

Comment:
What is the difference between this `keysKey` and `valuesKey`?

Reply:
We want to store key-value pairs like in SharedPreferences, so we want to encrypt either keys and values :) For some reason, we cannot use `AES` for keys encryption

Reply:
XOR its not an encryption it is just transformation. But what is difference between this line
`.keyEncryption(XorKeyEncryption(keysKey))` and next one. Why the lib requires two keys for AES encryption?

Reply:
`XOR` is used to transform keys.
`AES` is used to transform values. We need to pass `key` and `iv` there